### PR TITLE
feat: vault health endpoint - (bad blocks, bad files)

### PR DESCRIPTION
**Pull request by TheSriram**
- This feature required a reverse mapping between blockid-> files
  - Cassandra now has a new columnfamily blockfiles

DEPENDS ON #261

### CODE REVIEW

**Comment by TheSriram** [comment]
This is not self healing yet, that functionality resides in #261

File: deuce/drivers/cassandra/cassandrametadatadriver.py [hunk 1]
```diff
@@ -70,6 +70,22 @@
     AND fileid = %(fileid)s
 '''
 
+CQL_GET_BAD_BLOCKS = '''
+    SELECT blockid
+    FROM blocks
+    WHERE projectid = %(projectid)s
+    AND vaultid = %(vaultid)s
+    AND isinvalid = true
```
Comment (raxyu) [inline deuce/drivers/cassandra/cassandrametadatadriver.py]: is 'true' a typo of True?

Comment (TheSriram) [inline deuce/drivers/cassandra/cassandrametadatadriver.py]: No, I was just being consistent with the code base.

**Comment by racklanta** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://ci.cldbckp.com/job/deuce-pullrequest/278/

**Comment by racklanta** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://ci.cldbckp.com/job/deuce-pullrequest/279/

File: deuce/transport/wsgi/v1_0/vaulthealth.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        """Returns the statistics on vault controller object"""
+        vault = Vault.get(vault_id)
+
+        if vault:
+            bad_blocks, bad_files = vault.get_vault_health()
+            response = {
+                'Vault': vault_id,
+                'Bad Blocks': bad_blocks,
+                'Bad Files': bad_files
+            }
```
Comment (BenjamenMeyer) [inline deuce/transport/wsgi/v1_0/vaulthealth.py]: Q: Should we include this information in the Vault Statistics?

Comment (TheSriram) [inline deuce/transport/wsgi/v1_0/vaulthealth.py]: Good point, I guess we can do that.

**Comment by BenjamenMeyer** [comment]
@TheSriram I'm not saying removing /vaults/{vaultid}/health, just that we should probably add it to the statistics too, if that's desirable.

File: deuce/transport/wsgi/v1_0/vaulthealth.py (hunk 2 above)
Comment (BenjamenMeyer) [inline deuce/transport/wsgi/v1_0/vaulthealth.py]: @TheSriram I'm not saying removing /vaults/{vaultid}/health, just that we should probably add it to the statistics too, if that's desirable.

Comment (TheSriram) [inline deuce/transport/wsgi/v1_0/vaulthealth.py]: It's done. :)

Comment (BenjamenMeyer) [inline deuce/transport/wsgi/v1_0/vaulthealth.py]: Cool.

File: deuce/transport/wsgi/v1_0/vault.py [hunk 3]
```diff
@@ -22,7 +22,13 @@ def on_get(self, req, resp, vault_id):
...
 
         if vault:
-            resp.body = json.dumps(vault.get_vault_statistics())
+            vault_stats = vault.get_vault_statistics()
+            bad_blocks, bad_files = vault.get_vault_health()
+
+            vault_stats['metadata']['blocks']['bad'] = bad_blocks
+            vault_stats['metadata']['files']['bad'] = bad_files
+
+            resp.body = json.dumps(vault_stats)
```
Comment (BenjamenMeyer) [inline deuce/transport/wsgi/v1_0/vault.py]: Please update the vault statistics tests too.

This should probably be done at the metadata driver level as that is what builds the statistics dictionary.

Comment (TheSriram) [inline deuce/transport/wsgi/v1_0/vault.py]: Done.

**Comment by racklanta** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://ci.cldbckp.com/job/deuce-pullrequest/281/

File: deuce/tests/test_sqlite_storage_driver.py [hunk 4]
```diff
@@ -44,16 +44,29 @@ def test_vault_statistics(self):
...
+
+        for block_id in block_ids:
+            driver.mark_block_as_bad(vault_id, block_id)
+
         new_stats = driver.get_vault_statistics(vault_id)
 
         self.assertEqual(new_stats['blocks']['count'], 3)
         self.assertEqual(new_stats['files']['count'], 1)
+        self.assertEqual(new_stats['blocks']['bad'], 3)
+        self.assertEqual(new_stats['files']['bad'], 0)
```
Comment (BenjamenMeyer) [inline deuce/tests/test_sqlite_storage_driver.py]: Should have a bad files case in here too

Comment (BenjamenMeyer) [inline deuce/tests/test_sqlite_storage_driver.py]: Would be nice to have a bad files case here too

File: deuce/transport/wsgi/v1_0/vault.py [hunk 5]
```diff
@@ -22,7 +22,13 @@ def on_get(self, req, resp, vault_id):
         vault = Vault.get(vault_id)
 
         if vault:
-            resp.body = json.dumps(vault.get_vault_statistics())
+            vault_stats = vault.get_vault_statistics()
+            # bad_blocks, bad_files = vault.get_vault_health()
+            #
+            # vault_stats['metadata']['blocks']['bad'] = bad_blocks
+            # vault_stats['metadata']['files']['bad'] = bad_files
```
Comment (BenjamenMeyer) [inline deuce/transport/wsgi/v1_0/vault.py]: Please remove the commented lines

Comment (TheSriram) [inline deuce/transport/wsgi/v1_0/vault.py]: Yeah, this should have been removed. sorry

**Comment by BenjamenMeyer** [comment]
:+1:

**Comment by BenjamenMeyer** [comment]
Travis Failued to do Filecat transient failure

**Comment by racklanta** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://ci.cldbckp.com/job/deuce-pullrequest/282/

**Comment by TheSriram** [comment]
fixing api tests.

**Comment by jc7998** [comment]
:+1: 
Remember to update Apiary

**Comment by racklanta** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://ci.cldbckp.com/job/deuce-pullrequest/285/

**Comment by TheSriram** [comment]
rebased.

**Comment by racklanta** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://ci.cldbckp.com/job/deuce-pullrequest/296/

**Comment by BenjamenMeyer** [comment]
:+1:

**Comment by raxyu** [comment]
:+1:

**Pull request merged**